### PR TITLE
AWS: Enable ICMP Type 3 Code 4 for API server ELBs

### LINKED ISSUE #5850: API inaccessible on AWS as ICMP is not allowed - this breaks Path MTU Discovery
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.10.0 (git-8b52ea6d1)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
 
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T18:02:47Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:05:37Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

**3. What cloud provider are you using?**

AWS - provisioned in eu-west-1 (Ireland)

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

This is the command from my bash script. The ssh key file location is passed in and I'm using the vpc/subnets from my aws account - subnets are private and the utility subnets are public.

kops create cluster \
	  --node-count 2 \
	  --node-size m3.medium \
	  --master-count 1 \
	  --master-size t2.medium \
	  --zones=eu-west-1a,eu-west-1b \
	  --cloud-labels Owner=International,Example=DoThis \
	  --vpc=vpc-b35ee0d4 \
	  --subnets=subnet-e1d169a8,subnet-4ebb4415 \
	  --utility-subnets=subnet-74d46c3d,subnet-6cba4537 \
	  --ssh-public-key $SSH_PUB_KEY_LOCATION \
	  --api-loadbalancer-type internal \
	  --associate-public-ip=false \
	  --dns private \
	  --topology private \
	  --networking amazon-vpc-routed-eni \
	  --cloud aws

**5. What happened after the commands executed?**

I am unable to hit the api endpoint at all. TLS handshaking fails - even if I just attempt a curl -v at it. The security group attached to the API load balance (ELB) only allows TCP 443 for ingress. If I add a new rule or add an additional security group which allows all ICMP, it immediately fixes the problem.

**6. What did you expect to happen?**

I expected that ICMP would be allowed.

I understand that ICMP is required to allow 'Path MTU Discovery' ( https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/network_mtu.html#path_mtu_discovery ).

'kube-aws' had this issue reported and fixed: https://github.com/kubernetes-incubator/kube-aws/issues/214

Looking at the kops source, it *looks* like the aws provider adds the ICMP rule here, but I'm not that comfortable with go so I'm not entirely sure if I'm reading this correctly:

https://github.com/kubernetes/kops/blob/3c22c2c01a4cf3c2408f7a69708923d7b58552b3/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws.go#L3508

Cluster manifest: https://gist.github.com/cannontrodder/caa906480cc685a421316666d0209a68

### PR DESCRIPTION
**Pull request by davidarcher**
Resolves #5850

This is already fixed in kubernetes aws cloudprovider integration:  https://github.com/kubernetes/kubernetes/pull/27677

### CODE REVIEW

File: pkg/model/awsmodel/api_loadbalancer.go [hunk 1]
```diff
@@ -200,7 +202,20 @@ func (b *APILoadBalancerBuilder) Build(c *fi.ModelBuilderContext) error {
 				Protocol:      s("tcp"),
 			}
 			c.AddTask(t)
-		}
+
+			// Allow ICMP traffic required for PMTU discovery
+			t := &awstasks.SecurityGroupRule{
```
[the remark this answers is not in the document]
Comment (davidarcher) [reply pkg/model/awsmodel/api_loadbalancer.go]: Alright, cool. I've just wrap-inlined both then.

**Comment by davidarcher** [comment]
@gambol99 I've got the TF updated with the new sec group rule and the CI is passing. Should be ready for another look. Thanks!

**Comment by chrisz100** [comment]
/lgtm 
👍

**Comment by adammw** [comment]
This change pushes us over the 50 rule limit for a security group - any ideas on how we would configure this or workaround the limit? Is there any security risk on enabling these ICMP packets to the wider internet?

**Comment by arunkumarmurugesan** [comment]
Yes. I'm also having the same concern why do we need the ICMP packet enabled to the wider internet? 

Also, whenever I create new ELB, It's enabling ICMP rule in the ELB Security Group. Why so?

**Comment by infa-sathkuma** [comment]
Most security scans flags the 0.0.0.0/0 ICMP also has a vulnerability, I don't think they look at the types.  it would be helpful if someone explains  any  use case where the loadbalancer would send a packet with larger MTU ?  or backend. as far as I know MSS is a negotiated value and that takes up large part of the packet and 40 bytes extra to form the MTU. just trying to understand if what scenario the the PMTU is needed. most of the cases would involve IPSec tunnels.

**Pull request merged**